consensus/src/chained_bft/liveness/multi_proposer_election.rs

```diff
@@ -73,10 +84,58 @@ impl<T: Payload> ProposerElection<T> for MultiProposer {
         self.get_candidates(round)
     }
 
-    fn process_proposal(&self, proposal: Block<T>) -> Option<Block<T>> {
-        // At this phase we're considering the primary proposer only.
-        if let Some(author) = proposal.author() {
-            if author == self.get_candidates(proposal.round())[0] {
+    fn process_proposal(&mut self, proposal: Block<T>) -> Option<Block<T>> {
+        let author = match proposal.author() {
+            Some(author) => author,
+            None => {
+                return None;
+            }
+        };
+        let round = proposal.round();
+        let mut proposal_wrap = Some(proposal); // to indicate whether we've used it for logging
+        let candidates = self.get_candidates(round);
+        for (rank, candidate) in candidates.iter().enumerate() {
+            if rank == 0 && author == *candidate {
+                // Primary proposal
+                let proposal = proposal_wrap.take().unwrap();
```

Comment:
I think we have guideline not to use `unwrap` but rather use `expect`
But also I think you can just rewrite this function without `proposal_wrap`